modules/core/src/test/java/org/apache/ignite/internal/client/thin/events/FakeIgniteServer.java

```diff
@@ -0,0 +1,261 @@
+import java.util.EnumSet;
+import java.util.UUID;
+import java.util.function.Consumer;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.binary.BinaryRawWriter;
+import org.apache.ignite.failure.FailureType;
+import org.apache.ignite.internal.binary.BinaryReaderExImpl;
+import org.apache.ignite.internal.binary.BinaryWriterExImpl;
+import org.apache.ignite.internal.binary.streams.BinaryByteBufferInputStream;
+import org.apache.ignite.internal.binary.streams.BinaryHeapOutputStream;
+import org.apache.ignite.internal.binary.streams.BinaryInputStream;
+import org.apache.ignite.internal.client.thin.ProtocolContext;
+import org.apache.ignite.internal.client.thin.ProtocolVersion;
+import org.apache.ignite.internal.client.thin.ProtocolVersionFeature;
+import org.apache.ignite.internal.client.thin.io.gridnioserver.GridNioClientParser;
+import org.apache.ignite.internal.processors.odbc.ClientListenerRequest;
+import org.apache.ignite.internal.processors.platform.client.ClientStatus;
+import org.apache.ignite.internal.util.nio.GridNioCodecFilter;
+import org.apache.ignite.internal.util.nio.GridNioServer;
+import org.apache.ignite.internal.util.nio.GridNioServerListener;
+import org.apache.ignite.internal.util.nio.GridNioSession;
+import org.apache.ignite.internal.util.nio.GridNioSessionMetaKey;
+import org.apache.ignite.testframework.junits.JUnitAssertAware;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A fake ignite server for testing handshake and connection errors handling on the thin client side.
+ */
+public class FakeIgniteServer extends JUnitAssertAware implements GridNioServerListener<ByteBuffer>, AutoCloseable {
+    /** */
+    static final int HANDSHAKE_PERFORMED = GridNioSessionMetaKey.nextUniqueKey();
+
+    /** */
+    public static final byte[] EMPTY_BYTES = new byte[0];
+
+    /** */
+    private final GridNioServer<ByteBuffer> srv;
+
+    /** */
+    private final EnumSet<ErrorType> errorTypes;
+
+    /** */
+    private final ProtocolVersion protoVer;
+
+    /** */
+    private final UUID nodeId = UUID.randomUUID();
+
+    /** */
+    public FakeIgniteServer(InetAddress addr, int port, IgniteLogger logger) {
+        this(addr, port, logger, null, null);
+    }
+
+    /** */
+    public FakeIgniteServer(InetAddress addr, int port, IgniteLogger logger, EnumSet<ErrorType> errorTypes) {
+        this(addr, port, logger, null, errorTypes);
+    }
+
+    /** */
+    public FakeIgniteServer(InetAddress addr, int port, IgniteLogger logger, ProtocolVersion protoVer) {
+        this(addr, port, logger, protoVer, null);
+    }
+
+    /** */
+    public FakeIgniteServer(
+        InetAddress addr,
+        int port,
+        IgniteLogger logger,
+        ProtocolVersion protoVer,
+        EnumSet<ErrorType> errorTypes
+    ) {
+        this.protoVer = protoVer != null ? protoVer : ProtocolVersion.LATEST_VER;
```

Comment:
Perhaps, it would be better to fix protocol version here (use 1.7.0 for example), to avoid fake server modification on each protocol change